staging/src/k8s.io/apimachinery/pkg/runtime/openapi/parser.go

```diff
@@ -0,0 +1,149 @@
+/*
+Copyright 2018 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package openapi
+
+import (
+	"fmt"
+
+	golangproto "github.com/golang/protobuf/proto"
+	openapi_v2 "github.com/googleapis/gnostic/OpenAPIv2"
+
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/kube-openapi/pkg/util/proto"
+)
+
+const (
+	// ContentTypePBv2 is the content type for openapi v2 in protobuf format.
+	// This should be the content type requested from the downloader so
+	// a parserPBv2 will be able to parse the spec.
+	ContentTypePBv2 = "application/com.github.proto-openapi.spec.v2@v1.0+protobuf"
+
+	// groupVersionKindExtensionKey is the key used to lookup the
+	// GroupVersionKind value for an object definition from the
+	// definition's "extensions" map.
+	groupVersionKindExtensionKey = "x-kubernetes-group-version-kind"
+)
+
+type parserPBv2 struct{}
```

Comment:
Have you look at this: https://github.com/kubernetes/kube-openapi/tree/master/pkg/util/proto

Just want to make sure you explored the idea of sharing this parser with the code there if there is any chance. I am not expert on that code and not even sure if the parser is there. If you are aware of that package, just ignore this comment.

Reply:
This is basically just a wrapper for that, which creates models using kube-openapi/pkg/util/proto.NewOpenAPIData(doc) and converts a schema.GroupVersionKind to the modelname accepted by kube-openapi/pkg/util/proto.LookupModel